File: ruby/lib/products/catalogue.rb
require_relative './sku'

module Products

	class Catalogue

		class << self

			def from_file(catalogueReader)
				products = catalogueReader.readAll.map { |line|
					Sku.new([ line[0], line[1], line[2] ], line[3], line[4], Integer(line[5]))
				}
				Catalogue.new(products)
			end

		end

		def initialize(products)
			@products = products
		end

		def list(out)
			sorted_skus.each do |sku|
				out.printf("%s\t%dp\t%s\n", sku.code[2], sku.price, sku.title)
			end
			out.puts
			out.puts "10% discount on orders over £20.00!"
		end

		def lookup(id)
			@products.each do |sku|											# TODO -- select
				return sku if sku.code[2] == id
			end
			nil
		end

		def lookup_code(sku)
			@products.each do |s|												# TODO -- select
				return s.code if s.code[2] == sku
			end
			nil
		end

		private

		def sorted_skus
			@products.sort
		end

	end
end

File: python3/products/catalogue.py
from . sku import Sku

class Catalogue:

    @staticmethod
    def fromFile(catalogueReader):
        products = list(map(lambda line: Sku([line[0], line[1], line[2]], line[3], line[4], int(line[5])), catalogueReader.readAll()))
        return Catalogue(products)

    def __init__(self, products):
        self.products = products

    def list(self, out):
        for sku in self._sortedSkus():
            out.write("%s\t%dp\t%s\n" % (sku.code[2], sku.price, sku.title))
        out.write("\n10% discount on orders over $20.00!\n")

    def lookup(self, id):
        for sku in self.products:
            if sku.code[2] == id: return sku
        return None

    def lookupCode(self, sku):
        for s in self.products:
            if s.code[2] == sku: return s.code
        return None

    def _sortedSkus(self):
        return sorted(self.products, key=lambda sku: sku.code)

